repo/content/content_manager_indexes.go

```diff
@@ -0,0 +1,380 @@
+	for _, indexBlob := range indexBlobs {
+		if indexBlob.BlobID == compactedIndexBlob {
+			formatLog(ctx).Debugf("compaction was a no-op")
+			return nil
+		}
+	}
+
+	logEntry.OutputBlobs = append(logEntry.OutputBlobs, compactedIndexBlob)
+
+	logEntryBytes, err := json.Marshal(logEntry)
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal log entry bytes")
+	}
+
+	compactionLogBlobID, err := bm.encryptAndWriteBlobNotLocked(ctx, logEntryBytes, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "unable to write compaction log")
+	}
+
+	formatLog(ctx).Debugf("compacted indexes %v into %v and wrote log %v in %v", logEntry.InputBlobs, logEntry.OutputBlobs, compactionLogBlobID, time.Since(t0)) // allow:no-inject-time
+
+	if err := bm.deleteOldIndexBlobs(ctx, compactionLogBlobID); err != nil {
+		return errors.Wrap(err, "error deleting old index blobs")
+	}
+
+	return nil
+}
+
+func (bm *lockFreeManager) getCompactionLogEntries(ctx context.Context, blobs []blob.Metadata) (map[blob.ID]*compactionLogEntry, error) {
+	results := map[blob.ID]*compactionLogEntry{}
+
+	for _, cb := range blobs {
+		data, err := bm.getIndexBlobInternal(ctx, cb.BlobID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to read compaction blob %q", cb.BlobID)
+		}
+
+		le := &compactionLogEntry{}
+
+		if err := json.Unmarshal(data, le); err != nil {
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (bm *Manager) deleteOldIndexBlobs(ctx context.Context, latestBlobID blob.ID) error {
+	allCompactionLogBlobs, err := bm.listCache.listIndexBlobs(ctx, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing compaction log blobs")
+	}
+
+	// look for server-assigned timestamp of the compaction log entry we just wrote as a reference.
+	// we're assuming server-generated timestamps are somewhat reasonable and time is moving
+	latestBlobTime := blobTime(allCompactionLogBlobs, latestBlobID)
+	if latestBlobTime.IsZero() {
+		formatLog(ctx).Warningf("compaction log blob %q was not found in list results due to eventual consistency, ignoring", latestBlobID)
+		return nil
+	}
+
+	fetchCompactionLogs := blobsOlderThan(allCompactionLogBlobs, latestBlobTime.Add(-blobDeleteSafeTime))
+
+	entries, err := bm.getCompactionLogEntries(ctx, fetchCompactionLogs)
+	if err != nil {
+		return errors.Wrap(err, "unable to get compaction log entries")
+	}
+
+	oldCompactedBlobs := map[blob.ID]bool{}
+
+	for _, cl := range entries {
+		for _, b := range cl.InputBlobs {
+			log(ctx).Debugf("will delete old index %q compacted to %v", b, cl.OutputBlobs)
+
+			oldCompactedBlobs[b] = true
+		}
+	}
+
+	for cb := range oldCompactedBlobs {
+		log(ctx).Debugf("deleting compacted blob %v", cb)
+
+		if err := bm.st.DeleteBlob(ctx, cb); err != nil && err != blob.ErrBlobNotFound {
+			formatLog(ctx).Warningf("unable to delete compacted blob %v, %v", cb, err)
```

Comment:
Doesn't this result in a case similar to the one this patch is intended to solve?
If the corresponding compaction log entry is removed, then this index pack becomes _"valid"_ again, and thus any content (not-marked as deleted) in this index pack becomes "live" again.

Reply:
An option here is to avoid deleting the corresponding compaction log for the index blob that could not be deleted.
Would that work?

Reply:
you're right

we need to delete indexes first, wait enough time for eventual consistency in lists to stop returning them, then delete logs.

If we delete them at the same time, it's entirely possible that log will not be returned by list() but index blob will

Reply:
Right. There's a need for an additional window between removing the index packs and the removal of the compaction logs.

This is 2-phase GC for compaction logs ;) 🐢 🐢 all the way down.